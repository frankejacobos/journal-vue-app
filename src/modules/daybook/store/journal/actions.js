import journalApi from "@/api/journalApi";

export const createEntry = async ({ commit }, entry) => {
  const { date, text, picture } = entry;
  const dataToSave = { date, text, picture };
  const { data } = await journalApi.post("/entries.json", dataToSave);
  commit("addEntry", { ...dataToSave, id: data.name });
  return data.name;
};

export const readEntries = async ({ commit }) => {
  const { data } = await journalApi.get("/entries.json");
  let entries = [];
  if (data) {
    for (let id of Object.keys(data)) {
      entries.push({
        id,
        ...data[id],
      });
    }
  }
  commit("setEntries", entries);
};

export const updateEntry = async ({ commit }, entry) => {
  const { date, text, picture } = entry;
  const dataToSave = { date, text, picture };
  await journalApi.put(`/entries/${entry.id}.json`, dataToSave);
  commit("updateEntry", { ...entry });
};

export const deleteEntry = async ({ commit }, id) => {
  await journalApi.delete(`/entries/${id}.json`);
  commit("deleteEntry", id);
};

export const clearEntries = ({ commit }) => {
  commit("clearEntries");
};
