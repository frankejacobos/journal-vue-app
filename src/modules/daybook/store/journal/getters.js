// export const getter = (state) => {
//   return state.property;
// };

export const getEntriesByTerm =
  (state) =>
  (term = "") => {
    return term.length > 0
      ? state.entries.filter((entry) =>
          entry.text.toLowerCase().includes(term.toLowerCase())
        )
      : state.entries;
  };

export const getEntryById =
  (state) =>
  (id = "") => {
    const entry = state.entries.find((entry) => entry.id === id);
    return entry ? { ...entry } : undefined;
  };
