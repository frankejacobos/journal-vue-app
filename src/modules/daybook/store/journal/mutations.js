export const setEntries = (state, entries) => {
  state.entries = [...state.entries, ...entries];
  state.isLoading = false;
};

export const addEntry = (state, entry) => {
  state.entries = [entry, ...state.entries];
};

export const updateEntry = (state, entry) => {
  const index = state.entries.findIndex((e) => e.id === entry.id);
  state.entries[index] = entry;
};

export const deleteEntry = (state, entryId) => {
  state.entries = state.entries.filter((e) => e.id !== entryId);
};
