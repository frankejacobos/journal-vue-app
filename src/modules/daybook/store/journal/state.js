const date1 = new Date();
date1.setDate(date1.getDate() - 15);
const date2 = new Date();
date2.setDate(date2.getDate() - 5);
const date3 = new Date();
date3.setDate(date3.getDate() - 1);

const dates = [date1, date2, date3];

export default () => ({
  isLoading: true,
  entries: [
    {
      id: "1",
      date: dates[0].toDateString(),
      picture: null,
      text: "Officia irure id excepteur labore. Reprehenderit esse aliqua nulla amet reprehenderit culpa aute dolore consequat aute consequat nostrud. Culpa id tempor sint voluptate anim reprehenderit fugiat.",
    },
    {
      id: "2",
      date: dates[1].toDateString(),
      picture: null,
      text: "Ut laborum aute proident laboris occaecat do quis nulla. Ut nostrud enim do ipsum ut aliqua reprehenderit ipsum labore adipisicing do mollit esse culpa. Minim occaecat non nostrud dolor incididunt enim. Sit quis elit sit esse est ipsum commodo eu sint do ipsum mollit laborum. Nulla ullamco est voluptate et nulla sunt.",
    },
    {
      id: "3",
      date: dates[2].toDateString(),
      picture: null,
      text: "Eiusmod commodo do nisi est pariatur. Id dolor fugiat ut nulla. Tempor veniam commodo cillum proident officia do pariatur proident eu culpa nostrud ex amet. Enim Lorem cupidatat laboris in irure Lorem id ex esse aliquip. Ea ullamco ut adipisicing dolore sunt sit aliquip cillum pariatur et adipisicing cupidatat cillum. Do eiusmod minim dolore esse.",
    },
  ],
});
