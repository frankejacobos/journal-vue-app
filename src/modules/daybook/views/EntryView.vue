<template>
  <template v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2">
      <div class="entry-title d-flex align-items-center">
        <span class="text-success fs-4 fw-bold">{{ day }}</span>
        <span class="mx-1 fs-5">{{ month }} {{ year }}</span>
      </div>
      <div>
        <input
          type="file"
          @change="onFileChange"
          ref="imageSelector"
          v-show="false"
          accept="image/*"
        />
        <button class="btn btn-danger mx-2" @click="onDelete">
          Borrar
          <i class="fa fa-trash-alt"></i>
        </button>
        <button class="btn btn-primary" @click="onSelectImage">
          Subir foto
          <i class="fa fa-upload"></i>
        </button>
      </div>
    </div>
    <hr />
    <div class="d-flex flex-column px-3 h-75">
      <textarea
        placeholder="¿Qué está pasando en el mundo 2022?"
        v-model="entry.text"
        class="form-control"
        rows="10"
      ></textarea>
    </div>
    <img
      v-if="entry.picture && !localImage"
      :src="entry.picture"
      alt="entry-picture"
      class="img-thumbnail"
    />
    <img
      v-if="localImage"
      :src="localImage"
      alt="entry-picture"
      class="img-thumbnail"
    />
    <fab-icon icon="fa-save" @on:click="onSave"></fab-icon>
  </template>
</template>

<script>
import { defineAsyncComponent } from "vue";
import { mapActions, mapGetters } from "vuex";
import getDayMonthYear from "../helpers/getDayMonthYear";
import Swal from "sweetalert2";
import uploadImage from "../helpers/uploadImage";

export default {
  name: "EntryView",
  props: {
    id: {
      type: String,
      required: true,
    },
  },
  components: {
    FabIcon: defineAsyncComponent(() => import("../components/FabIcon.vue")),
  },
  data() {
    return {
      entry: null,
      localImage: null,
      file: null,
    };
  },
  computed: {
    ...mapGetters("journal", ["getEntryById"]),
    day() {
      const { day } = getDayMonthYear(this.entry.date);
      return day;
    },
    month() {
      const { month } = getDayMonthYear(this.entry.date);
      return month;
    },
    year() {
      const { year } = getDayMonthYear(this.entry.date);
      return year;
    },
  },
  methods: {
    ...mapActions("journal", ["createEntry", "updateEntry", "deleteEntry"]),
    getEntry() {
      const entry = this.getEntryById(this.id);
      if (this.id === "new") {
        this.entry = {
          text: "",
          date: new Date().getTime(),
        };
      } else {
        if (!entry) return this.$router.push({ name: "no-entry" });
        this.entry = entry;
      }
    },
    async onSave() {
      new Swal({
        title: "Espere",
        text: "Guardando información",
        icon: "info",
        allowOutsideClick: false,
      });
      Swal.showLoading();
      const picture = await uploadImage(this.file);
      this.entry.picture = picture;
      if (this.entry.id) {
        await this.updateEntry(this.entry);
      } else {
        const index = await this.createEntry(this.entry);
        this.$router.push({ name: "entry", params: { id: index } });
      }
      Swal.fire({
        title: "Guardado",
        text: "Se guardó correctamente",
        icon: "success",
      });
      this.localImage = null;
    },
    async onDelete() {
      const { isConfirmed } = await Swal.fire({
        title: "¿Estás seguro?",
        text: "No podrás revertir esta acción",
        icon: "warning",
        showCancelButton: true,
        confirmButtonText: "Sí, borrar",
        cancelButtonText: "Cancelar",
      });
      if (isConfirmed) {
        Swal.fire({
          title: "Espere",
          text: "Borrando información",
          icon: "info",
          allowOutsideClick: false,
        });
        Swal.showLoading();
        await this.deleteEntry(this.id);
        this.$router.push({ name: "no-entry" });
        Swal.fire({
          title: "Borrado",
          text: "Se borró correctamente",
          icon: "success",
        });
      }
    },
    onFileChange(e) {
      const file = e.target.files[0];
      if (!file) {
        this.localImage = null;
        this.file = null;
        return;
      }
      this.file = file;
      const reader = new FileReader();
      reader.onload = (e) => {
        this.localImage = e.target.result;
      };
      reader.readAsDataURL(file);
    },
    onSelectImage() {
      this.$refs.imageSelector.click();
    },
  },
  created() {
    this.getEntry();
  },
  watch: {
    id() {
      this.getEntry();
    },
  },
};
</script>

<style lang="scss" scoped>
textarea {
  font-size: 20px;
  border: none;
  height: 100%;
  &:focus {
    outline: none;
  }
}
img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0px 5px 10px rgba(0, 0, 0, 0.5);
}
</style>
