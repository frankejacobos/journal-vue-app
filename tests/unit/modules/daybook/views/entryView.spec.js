import { createStore } from "vuex";
import { shallowMount } from "@vue/test-utils";
import { journalState } from "@/../tests/unit/mocks/test-journal-state";
import journal from "@/modules/daybook/store/journal";
import EntryView from "@/modules/daybook/views/EntryView.vue";
import Swal from "sweetalert2";

jest.mock("sweetalert2", () => ({
  fire: jest.fn(),
  showLoading: jest.fn(),
  close: jest.fn(),
}));
const createVuexStore = (state) =>
  createStore({
    modules: {
      journal: {
        ...journal,
        state: { ...state },
      },
    },
  });
const store = createVuexStore(journalState);
store.dispatch = jest.fn();
const mockRouter = { push: jest.fn() };
let wrapper;
describe("Pruebas en el componente EntryView", () => {
  const exec = (id) => {
    wrapper = shallowMount(EntryView, {
      props: { id },
      global: {
        mocks: { $router: mockRouter },
        plugins: [store],
      },
    });
  };
  afterEach(() => {
    mockRouter.push.mockClear(); /* limpiar mocks */
  });
  test("debe sacar al usuario si el id no existe", () => {
    exec("3");
    expect(mockRouter.push).toHaveBeenCalledWith({ name: "no-entry" });
  });
  test("debe de mostrar el entry si el id existe", () => {
    exec("1");
    expect(wrapper.html()).toMatchSnapshot();
    expect(mockRouter.push).not.toHaveBeenCalled();
  });
  test("debe de llamar al deleteEntry al hacer click en el botón", (done) => {
    exec("1");
    Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true }));
    wrapper.find(".btn-danger").trigger("click");
    expect(Swal.fire).toHaveBeenCalledWith({
      title: "¿Estás seguro?",
      text: "No podrás revertir esta acción",
      icon: "warning",
      showCancelButton: true,
      confirmButtonText: "Sí, borrar",
      cancelButtonText: "Cancelar",
    });
    setTimeout(() => {
      expect(store.dispatch).toHaveBeenCalledWith("journal/deleteEntry", "1");
      expect(mockRouter.push).toHaveBeenCalledWith({ name: "no-entry" });
      done();
    }, 1);
  });
});
