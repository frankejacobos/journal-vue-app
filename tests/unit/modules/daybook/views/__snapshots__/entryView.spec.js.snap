// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente EntryView debe de mostrar el entry si el id existe 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div class="entry-title d-flex align-items-center"><span class="text-success fs-4 fw-bold">1</span><span class="mx-1 fs-5">Enero 2020, Miércoles</span></div>
  <div><input type="file" accept="image/*" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué está pasando en el mundo 2022?" class="form-control" rows="10"></textarea></div>
<!--v-if-->
<!--v-if-->
<fab-icon-stub icon="fa-save"></fab-icon-stub>
`;
